src/file.cr

```diff
@@ -412,7 +412,8 @@ class File < IO::FileDescriptor
   #   * `"c*"` matches all files beginning with `c`.
   #   * `"*c"` matches all files ending with `c`.
   #   * `"*c*"` matches all files that have `c` in them (including at the beginning or end).
-  # * `**` matches an unlimited number of arbitrary characters including `/`.
+  # * `**` matches directories recursively if followed by `/`.
+  #   If this path segment contains any other characters, it is the same as the usual `*`.
```

Comment:
Could you give an example of what this means? I know this came from Ruby, but I still don't understand it

Reply:
Oops, Okay, maybe it should be describe better than ruby version.

let us use crystal-book project as root, i write a new `1.cr`, with following content:

```crystal
# 1. If we are use ** without followed by a /, or followed by others, e.g. followed by `.md`, it same as `*`

p Dir.glob("docs/**.md")
p Dir.glob("docs/*.md") # totally same as above.

# Dir.glob("docs/**.md") # => ["docs/platform_support.md", "docs/SUMMARY.md", "docs/governance.md", "docs/README.md"]
# Dir.glob("docs/*.md") # => ["docs/platform_support.md", "docs/SUMMARY.md", "docs/governance.md", "docs/README.md"]

# 2. But if ** followed by a `/`, it matches directories recursively. (behavior not same as `*`)

x =  Dir.glob("docs/**/*.md") # => ["docs/guides/ci/circleci.md", "docs/guides/ci/gh-actions.md", "docs/guides/ci/README.md",
# ...... a very long list which include all md file in any depth subdirectory ...
# "do "docs/database/connection.md", "docs/database/README.md", "docs/database/transactions.md"]
p x.size # => 143

y =  Dir.glob("docs/*/*.md") # ["docs/conventions/coding_style.md", "docs/database/README.md",
# .... a long list which include all md file in level 1 and level 2 subdirectory of docs/.
# , "docs/syntax_and_semantics/while.md", "docs/the_shards_command/README.md", "docs/using_the_compiler/README.md"]
p y.size # => 92
```

Reply:
Thanks!

I'm thinking that `**.md` should actually produce a runtime error. It doesn't make sense to me if it's the same as `*`... 🤔

Reply:
> I'm thinking that **.md should actually produce a runtime error.

Yes, i think so, runtime error is better, in fact, i never use like `Dir.glob("docs/**.md")` before.

Reply:
No, we should not change established semantics. The glob patterns are not our own invention, they work the same way as in bash and multiple other shells.
`ls docs/**.md` lists exactly the same paths as `Dir.glob("docs/**.md")`.

This is completely valid though. `**` only has a special meaning as a recursive match when it's followed by a path separator (`/`). But with out that, it's just consecutive wildcards, each matching a string of any length. That's completely valid syntax and has a clear meaning. It doesn't make a semantic difference to using a single wildcard. But that's not a reason to make it a syntax error. Especially when the same thing works in all other glob implementations.

Reply:
I guess the first thought is, we don't like duplicate, if `docs/*.md` work same as `docs/**.md`, why we let user select?

 talking about convention, i think the really big issue(for newcomer) is the `"` and `'`, when migrate one Ruby project to Crystal, probably the most annoy(boring) work is change `'` to `"`.  for least surprised, i still propose make double and single quotes function same in 2.0?, we can introduce `?A`(ruby) or `$'A'`(bash) to present the `Char` type.